src/Io/Transaction.php

```diff
@@ -255,25 +258,37 @@ private function onResponseRedirect(ResponseInterface $response, RequestInterfac
     /**
      * @param RequestInterface $request
      * @param UriInterface $location
+     * @param int $statusCode
      * @return RequestInterface
+     * @throws \RuntimeException
      */
-    private function makeRedirectRequest(RequestInterface $request, UriInterface $location)
+    private function makeRedirectRequest(RequestInterface $request, UriInterface $location, $statusCode)
     {
         $originalHost = $request->getUri()->getHost();
-        $request = $request
-            ->withoutHeader('Host')
-            ->withoutHeader('Content-Type')
-            ->withoutHeader('Content-Length');
+        $request = $request->withoutHeader('Host')->withUri($location);
+
+        if (!$this->isTemporaryOrPermanentRedirect($statusCode)) {
+            $request = $request
+                ->withoutHeader('Content-Type')
+                ->withoutHeader('Content-Length');
+        }
 
         // Remove authorization if changing hostnames (but not if just changing ports or protocols).
         if ($location->getHost() !== $originalHost) {
             $request = $request->withoutHeader('Authorization');
         }
```

Comment:
```suggestion
        // Remove authorization if changing hostnames (but not if just changing ports or protocols).
        $originalHost = $request->getUri()->getHost();
        if ($location->getHost() !== $originalHost) {
            $request = $request->withoutHeader('Authorization');
        }
```
Part A2: This is the result from A1.